staging/src/k8s.io/cloud-provider/volume/helpers/rounding.go

```diff
@@ -41,139 +43,110 @@ const (
 // int32 instead of an int64 and an error if there's overflow
 func RoundUpToGiBInt32(size resource.Quantity) (int32, error) {
-	requestBytes, ok := size.AsInt64()
-	if !ok {
-		return 0, fmt.Errorf("quantity %s is too great, overflows int64", size.String())
-	}
-	return roundUpSizeInt32(requestBytes, GiB)
+	return roundUpSizeInt32(size, GiB)
 }
 
 // roundUpSizeInt calculates how many allocation units are needed to accommodate
-// a volume of given size. It returns an int instead of an int64 and an error if
-// there's overflow
-func roundUpSizeInt(volumeSizeBytes int64, allocationUnitBytes int64) (int, error) {
-	roundedUp := roundUpSize(volumeSizeBytes, allocationUnitBytes)
-	roundedUpInt := int(roundedUp)
-	if int64(roundedUpInt) != roundedUp {
-		return 0, fmt.Errorf("capacity %v is too great, casting results in integer overflow", roundedUp)
+// a volume of a given size. It returns an int and an error if there's overflow
+func roundUpSizeInt(size resource.Quantity, allocationUnitBytes int64) (int, error) {
+	if bits.UintSize == 32 {
+		res, err := roundUpSizeInt32(size, allocationUnitBytes)
+		return int(res), err
 	}
-	return roundedUpInt, nil
+	res, err := roundUpSizeInt64(size, allocationUnitBytes)
+	return int(res), err
 }
 
 // roundUpSizeInt32 calculates how many allocation units are needed to accommodate
-// a volume of given size. It returns an int32 instead of an int64 and an error if
-// there's overflow
-func roundUpSizeInt32(volumeSizeBytes int64, allocationUnitBytes int64) (int32, error) {
-	roundedUp := roundUpSize(volumeSizeBytes, allocationUnitBytes)
-	roundedUpInt32 := int32(roundedUp)
-	if int64(roundedUpInt32) != roundedUp {
-		return 0, fmt.Errorf("quantity %v is too great, overflows int32", roundedUp)
+// a volume of a given size. It returns an int32 and an error if there's overflow
+func roundUpSizeInt32(size resource.Quantity, allocationUnitBytes int64) (int32, error) {
+	roundedUpInt32, err := roundUpSizeInt64(size, allocationUnitBytes)
+	if err != nil {
+		return 0, err
 	}
-	return roundedUpInt32, nil
+	if roundedUpInt32 > math.MaxInt32 {
+		return 0, fmt.Errorf("quantity %s is too great, overflows int32", size.String())
+	}
+	return int32(roundedUpInt32), nil
 }
 
-// roundUpSize calculates how many allocation units are needed to accommodate
-// a volume of given size. E.g. when user wants 1500MiB volume, while AWS EBS
+// roundUpSize64 calculates how many allocation units are needed to accommodate
+// a volume of a given size. E.g. when user wants 1500MiB volume, while AWS EBS
 // allocates volumes in gibibyte-sized chunks,
-// RoundUpSize(1500 * 1024*1024, 1024*1024*1024) returns '2'
+// roundUpSizeInt64(1500MiB, 1024*1024*1024) returns '2'
 // (2 GiB is the smallest allocatable volume that can hold 1500MiB)
-func roundUpSize(volumeSizeBytes int64, allocationUnitBytes int64) int64 {
+// It returns an int64 and an error if the size was too great to work with
```

Comment:
Could you add a comment here or somewhere visible in this package summarizing some of the findings/speculation discussed in this PR and in the issue? This will be like round 3 of us trying to fix our rounding woes so it would be great to document for future readers what we are doing (rounding while checking for overflow) and why (different provider apis expect different scales(KB vs MB)/types(int vs int64)). It might also be worth mentioning that the resource.quantity library doesn't totally on its own solve this need because of the binary SI vs decimal SI scaling issue [although that raises the question of why we don't augment that I guess because that's a stable API that we can't change much whereas this is an internal API used by volume plugins.] )